host-apis/wasi-0.2.0/host_api.cpp

```diff
@@ -357,88 +344,50 @@ Result<HttpHeaders*> HttpHeaders::FromEntries(HttpHeadersGuard guard,
   wasi_http_0_2_0_types_header_error_t err;
   if (!wasi_http_0_2_0_types_static_fields_from_list(&tuples, &ret, &err)) {
     // TODO: handle `err`
-    return Result<HttpHeaders*>::err(154);
+    return Result<HttpHeaders *>::err(154);
   }
 
-  auto headers = new HttpHeaders(guard, std::make_unique<WASIHandle<HttpHeaders>>(ret));
-  return Result<HttpHeaders*>::ok(headers);
+  auto headers = new HttpHeaders(std::make_unique<WASIHandle<HttpHeaders>>(ret));
+  return Result<HttpHeaders *>::ok(headers);
 }
 
-HttpHeaders::HttpHeaders(HttpHeadersGuard guard, const HttpHeadersReadOnly &headers)
-  : HttpHeadersReadOnly(nullptr), guard_(guard) {
+HttpHeaders::HttpHeaders(const HttpHeadersReadOnly &headers) : HttpHeadersReadOnly(nullptr) {
   Borrow<HttpHeaders> borrow(headers.handle_state_.get());
   auto handle = wasi_http_0_2_0_types_method_fields_clone(borrow);
   this->handle_state_ = std::unique_ptr<HandleState>(new WASIHandle<HttpHeaders>(handle));
 }
 
 // We currently only guard against a single request header, instead of the full list in
 // https://fetch.spec.whatwg.org/#forbidden-request-header.
-static std::list forbidden_request_headers = {
-  "host",
+static std::vector forbidden_request_headers = {
+    "host",
 };
 
 // We currently only guard against a single response header, instead of the full list in
 // https://fetch.spec.whatwg.org/#forbidden-request-header.
-static std::list forbidden_response_headers = {
-  "host",
+static std::vector forbidden_response_headers = {
+    "host",
 };
 
-bool HttpHeaders::check_guard(HttpHeadersGuard guard, string_view header_name) {
-  std::list<const char*>* forbidden_headers = nullptr;
-  switch (guard) {
-  case HttpHeadersGuard::None:
-    return true;
-  case HttpHeadersGuard::Request:
-    forbidden_headers = &forbidden_request_headers;
-    break;
-  case HttpHeadersGuard::Response:
-    forbidden_headers = &forbidden_response_headers;
-    break;
-  default:
-    MOZ_ASSERT_UNREACHABLE();
-  }
-
-  if (!forbidden_headers) {
-    return true;
-  }
-
-  for (auto header : *forbidden_headers) {
-    if (header_name.compare(header) == 0) {
-      return false;
-    }
+std::vector<string_view> HttpHeaders::get_forbidden_request_headers() {
```

Comment:
Would this be better as a `const&` to ensure the vec itself isn't copied, while it's also not mutated?

Reply:
It is mean't to be mutated so we can take ownership of this list and then ensure it is properly sorted.

So the only change would be just be moving the clone into the caller, which I'm not sure is necessary?

Reply:
Added something along these lines.